constraint-solver/src/solver/mod.rs

```diff
@@ -0,0 +1,225 @@
+mod tests;
+mod utils;
+
+/// Current state of a variable in the solver.
+#[derive(Debug, PartialEq)]
+enum VariableState<T: FieldElement, V> {
+    /// The variable is unknown (but has a range constraint).
+    Unknown(RangeConstraint<T>),
+    /// The variable is known, and can be computed from other variables using
+    /// the given expression.
+    /// Note that if the expression is an instance of `SymbolicExpression::Concrete(_)`,
+    /// we have found a concrete variable assignment.
+    Known(SymbolicExpression<T, V>),
+}
+
+// We could derive it, but then we'd have to require that `V: Default`.
+impl<T: FieldElement, V> Default for VariableState<T, V> {
+    fn default() -> Self {
+        VariableState::Unknown(RangeConstraint::default())
+    }
+}
+
+impl<T: FieldElement, V> VariableState<T, V> {
+    fn range_constraint(&self) -> RangeConstraint<T> {
+        match self {
+            VariableState::Unknown(range_constraint) => range_constraint.clone(),
+            // Note that the expression's range constraint might not be as tight as it
+            // could be. But for solving, we only need range constraints of unknown variables,
+            // so this is fine.
+            VariableState::Known(expr) => expr.range_constraint(),
+        }
+    }
+}
+
+/// Given a list of constraints, tries to derive as many variable assignments as possible.
+pub struct Solver<T: FieldElement, V> {
+    /// The algebraic constraints to solve.
+    /// Note that these are mutated during the solving process.
+    /// They must be kept consistent with the variable states.
+    algebraic_constraints: Vec<QuadraticSymbolicExpression<T, V>>,
+    /// The current state of the variables.
+    variable_states: BTreeMap<V, VariableState<T, V>>,
+    /// Input nodes to the constraint system.
+    input_variables: BTreeSet<V>,
+}
+
+impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display + Debug> Solver<T, V> {
+    #[allow(dead_code)]
+    pub fn new(algebraic_constraints: Vec<QuadraticSymbolicExpression<T, V>>) -> Self {
+        let input_variables = known_variables(&algebraic_constraints);
+        let variable_states = input_variables
+            .iter()
+            .map(|known_var| {
+                (
+                    known_var.clone(),
+                    VariableState::Known(SymbolicExpression::Symbol(
+                        known_var.clone(),
+                        // TODO: We're losing the initial range constraint here!
+                        RangeConstraint::default(),
+                    )),
+                )
+            })
+            .collect();
+
+        Solver {
+            algebraic_constraints,
+            variable_states,
+            input_variables,
+        }
+    }
+
+    /// Solves the constraints as far as possible. For each variable it was able to
+    /// find a unique solution for, the result contains an expression, indicating
+    /// how the variable can be computed from other variables at runtime.
```

Comment:
I think we should be careful about taking the "runtime solving" view before we have a clear picture what the actual task is the solver is solving. I.e. I think we should first figure out what it should do before defining interfaces. One way could be to connect it to the rest of the system and see what can (and has to be) done.

For example, if two variables are unknown, but we can determine that they are equivalent, this would be very useful for the optimizer but this engine could not accomplish it like that.

Also it does simplify QSEs but does not return the simplified versions - why is that?